### PR TITLE
[WIP] Introduce form_with #25197

### PR DESCRIPTION
**Pull request by marekkirejczyk**
### Introducing form_with

Introduce form_with in place of form_for and form_tag.
This is a prototype and the purpose of this PR is to gather feedback.

### CODE REVIEW

**Comment by kaspth** [comment]
Thanks for making headway on this! ❤️

I'm going to try to review it in the coming days. Since this PR potentially has a large footprint let's keep it to just @dhh's original examples. Then we get a foundation to build on in other PRs. Thoughts?

File: actionview/lib/action_view/helpers/form_with_helper.rb [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+          result = (blank ? tag.option("", value: "") : "").html_safe
+          result += choices.map do |choice|
+            if choice.is_a? Array
+              tag.option choice[0], value: choice[1]
+            else
+              tag.option choice, value: choice
+            end
+          end.join("\n").html_safe
+          result
+        end
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: We can spare the `is_a?` check and handle the array-ness more cleanly with:

```ruby
def option_tags_for_select(choices, blank: false)
  built_options = choices.map do |(title, value)|
    tag.option title, value: value || title
  end.join("\n")

  if blank
    tag.option("", value: "").html_safe + built_options
  else
    built_options.html_safe
  end
end
```

Tested in the console with:

```ruby
irb(main):011:0> [ [ :horse, :cheese ] ].map { |(choice, value)| puts choice; puts value || choice }
horse
cheese
=> [nil]
irb(main):012:0> [ [ :horse ] ].map { |(choice, value)| puts choice; puts value || choice }
horse
horse
=> [nil]
irb(main):013:0> [ [ :horse, :cheese, :grill ] ].map { |(choice, value)| puts choice; puts value || choice }
horse
cheese
=> [nil]
```

File: actionview/lib/action_view/helpers/form_with_helper.rb [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+            hidden = tag.input(hidden_options).html_safe
+          end
+          hidden + tag.input(checkbox_options).html_safe
+        end
+
+        def select(method, choices = nil, blank: nil, prompt: nil, index: :undefined, disabled: nil, **options, &block)
+          tag.select option_tags_for_select(choices, blank: blank), options_for(method, options)
+        end
+
+        def collection_select(method, collection, value_method, text_method, blank: nil, prompt: nil, index: :undefined, disabled: nil, **options)
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: We should just kill this method entirely and just have `select`.

File: actionview/lib/action_view/helpers/form_with_helper.rb [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+              if value.respond_to?(:include?)
+                value.include?(on_value)
+              else
+                value.to_i == on_value.to_i
+              end
+            end
+          end
+
+          def translate_with_human_attribute_name(method)
+            model && model.class.respond_to?(:human_attribute_name) ? model.class.human_attribute_name(method) : nil
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: Don't think saving two lines here is worth it (plus I dislike the nil, since that's the default return value for a method) :

```ruby
if model && model.class.respond_to?(:human_attribute_name)
  model.class.human_attribute_name(method)
end
```

File: actionview/lib/action_view/helpers/form_with_helper.rb [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+              model_name = model.respond_to?(:model_name) ? model.model_name.human : scope.to_s.humanize
+              defaults = [:"helpers.submit.#{scope}.#{key}", :"helpers.submit.#{key}", :"#{key.to_s.humanize} #{model_name}"]
+              I18n.t(defaults.shift, model: model_name, default: defaults)
+            else
+              I18n.t("helpers.submit.no_model")
+            end
+          end
+
+          def name_for(method, options)
+            scope = options.delete(:scope) { @scope }
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: Is this to support changing the scope on just one field, e.g. `form.text_field :title, scope: :comment`?

Personally I'd sway more to tackle on some extra lines, like we do in routes.rb, than try to cram everything in to one line. That's how we end up with tons of options again.

I'd go with `fields(:permission) do |fields|` to change the scope.

What do you think, @dhh?

File: actionview/lib/action_view/helpers/form_with_helper.rb [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+            index = options.delete(:index)
+            method = method.to_s.chomp("?")
+            name = scope.to_s
+            name += index ? "[#{index}]" : ""
+            name += scope.nil? ? method : "[#{method}]"
+            name += options.delete(:multiple) ? "[]" : "" #exception => file_field
+          end
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: How about? 😁

```ruby
def input_name(attribute, index: nil, multiple: nil, **options)
  attribute = attribute.to_s.chomp("?")

  indexing   = "[#{index}]" if index
  multipling = "[]" if multiple

  if @scope
    "#{@scope}#{indexing}[#{attribute}]#{multipling}"
  else
    "#{attribute}#{indexing}#{multipling}"
  end
end
```

File: actionview/lib/action_view/helpers/form_with_helper.rb [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+          def id_for(method, scope, **options)
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: This doesn't seem to be used? Besides I think `dom_id(model)` is what's to be used for this.

File: actionview/lib/action_view/helpers/form_with_helper.rb [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+            type = selector.split("_").first
+            options = options_for(method, options)
+            options.merge!(value: model.send(method)) if model && type != 'password'
+            options.merge!(value: args[0]) if args.size > 0
+            options[:size] = options[:maxlength] unless options.key?(:size)
+            if placeholder = placeholder(options.delete(:placeholder), method)
+              options.merge!(placeholder: placeholder)
+            end
+            options.reverse_merge!(type: type)
+          end
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: How about this? 😁

```ruby
def #{selector}(method, *args, **options)
  tag.input options_for_field('#{selector.split("_").first}', method, args, options)
end 

def field_options(field_type, attribute, value = nil, *args, placeholder: nil, **options)
  options = default_options(attribute, options)

  if value
    options[:value] ||= value
  elsif @model && field_type.match?('password')
    options[:value] ||= @model.public_send(attribute)
  end

  options[:placeholder] = placeholder(placeholder, attribute)
  options[:size] ||= options[:max_length]
  options[:type] ||= field_type
  options
end
```

This pulls up the selector splitting, because we only need to do that once when we build the method.

Avoids `merge` calls because that allocates extra hashes and just assigning the key is clearer.

File: actionview/lib/action_view/helpers/form_with_helper.rb [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+                        :number_field, :range_field]
+
+        GENERATED_FIELD_HELPERS = FIELD_HELPERS - [:label, :check_box, :radio_button, :fields_for, :file_field, :text_area]
+
+        attr_accessor :model, :scope, :url, :remote
+
+        GENERATED_FIELD_HELPERS.each do |selector|
+          class_eval <<-RUBY_EVAL, __FILE__, __LINE__ + 1
+            def #{selector}(method, *args, **options)                             # def text_field(method, *args, **options)
+              tag.input options_for_field('#{selector}', method, args, options)   #   tag.input options_for_field(selector, method, args, options)
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: Can we pull this string outside the `class_eval`? I think we could save a needless string allocation that happens on each call to one of these helpers.

File: actionview/lib/action_view/helpers/form_with_helper.rb [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+        def label(method, *args, **options, &block)
+          content ||= args[0] if args.size > 0
+          content ||= I18n.t("#{model.model_name.i18n_key}.#{method}", default: "", scope: "helpers.label").presence if model && model.model_name
+          content ||= translate_with_human_attribute_name(method)
+          content ||= method.to_s.humanize
+          content = @template.capture(content, &block) if block_given?
+          tag.label content, options
+        end
+
+        def check_box(method, include_hidden: true, on: "1", off: "0", **options)
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: `method` confuses me because I keep thinking of the `_method` input field as well as the HTTP methods.

To me it makes much more sense to say `attribute` since that what we're working on.

**Comment by kaspth** [comment]
Also worth checking out if you haven't already: https://github.com/rails/rails/blob/b326e82dc012d81e9698cb1f402502af1788c1e9/actionview/lib/action_view/helpers/tags/base.rb 😁

I'll try to come back tomorrow with some more review! But just based on this quick dip building all these tags seems a lot more 😱 than I first thought. Lots of options n' stuff! All the more reason I think we should not try to fix the whole thing in this PR. I think we might even be cramming a bit too many fields in here already 😉

Rest assured, we'll find a way through it ❤️

File: actionview/lib/action_view/helpers/form_with_helper.rb (hunk 4 above)
Comment (marekkirejczyk) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: Yes this to support changing the scope as in your example. In defence of `scope` often times you do have fields like non related to the model, that would be useful to put into another scope, i.e. checkbox with terms and conditions. But I don't insist :)

File: actionview/lib/action_view/helpers/form_with_helper.rb (hunk 1 above)
Comment (bentbutnotbroken7469) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: Horse cock

Comment (bentbutnotbroken7469) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: If a horse has a little wang, do they call him people cock?

File: actionview/lib/action_view/helpers/form_with_helper.rb (hunk 7 above)
Comment (bentbutnotbroken7469) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: I think it's overkill a bit but hey stranger things have oroven usefill

Comment (bentbutnotbroken7469) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: 32769

File: actionview/lib/action_view/helpers/form_with_helper.rb [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+            options.merge!(value: model.send(method)) if model && type != 'password'
+            options.merge!(value: args[0]) if args.size > 0
+            options[:size] = options[:maxlength] unless options.key?(:size)
+            if placeholder = placeholder(options.delete(:placeholder), method)
+              options.merge!(placeholder: placeholder)
+            end
+            options.reverse_merge!(type: type)
+          end
+
+          def placeholder(tag_value, method)
```
Comment (bentbutnotbroken7469) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: Placeholder 76, here ya go I was just keepin that ass worm while you were in the bathroom

File: actionview/test/template/form_with_helper_test.rb [hunk 11]
```diff
@@ -0,0 +1,559 @@
...
+    assert_tag_equal('<input multiple="multiple" name="custom" type="file">') { |f| f.file_field(:blah, multiple: true, name: "custom") }
+    assert_tag_equal('<input multiple="multiple" name="custom[]" type="file">') { |f| f.file_field(:custom, multiple: true, scope: nil) }
+  end
+
+  def test_hidden_field
+    assert_tag_equal('<input name="post[title]" type="hidden" value="Catch 22">') { |f| f.hidden_field("title") }
+    assert_tag_equal('<input name="post[secret]" type="hidden" value="1">') { |f| f.hidden_field("secret?") }
+  end
+
+  def test_hidden_field_with_escapes
```
Comment (bentbutnotbroken7469) [inline actionview/test/template/form_with_helper_test.rb]: Open all hidden files in the world on Oct 28, for 2 mins, estimated increase in the monthy divirce rate, 35-40%

File: actionview/lib/action_view/helpers/form_with_helper.rb (hunk 4 above)
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: I don't like that scope here just obliterates the existing scope passed in to the form — shouldn't it add to the existing scope?

I mean, if you want the field in a completely different scope then I think we should encourage the use of `tag`. E.g.:

```ruby
form_with model: @post do |form|
  # Other fields...
  tag.input type: 'check_box', name: 'terms[conditions]', checked: false
end
```

Although that's a bit wordy, we could look into replacing `check_box_tag` and friends with `tag.check_box` and such.

File: actionview/lib/action_view/helpers/form_with_helper.rb [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+              false
+            when String
+              value == on_value
+            else
+              if value.respond_to?(:include?)
+                value.include?(on_value)
+              else
+                value.to_i == on_value.to_i
+              end
+            end
```
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: Let's see if we can't reuse all the tag helpers instead of copying the code over.

Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: On the other hand, it's also nice to be free from the old code to explore new ways of doing things. I don't think having everything in this file will work well.

Comment (marekkirejczyk) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: I would avoid using tag helpers for couple of reasons:
- They are not HTML5 compliant
- They have tons of deprecated logic e.g. retrieving object by name
- It seems to me they might be deprecated at some point as they main purpose seems to be implementation of form_for
- there is relatively little code to copy

Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: Great 👍, I just reached the same conclusion looking more into the code.

But let's not just copy code then, let's question the embedded assumptions.

Comment (marekkirejczyk) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: Each case is covered with test corresponding test. However I am not sure if we need all the different cases.

I think it would be easier and more efficient if you start code review from tests, so that we discuss interface and different variants & cases -> then I implement conclusions -> and the we go back to implementation details do the review and implement changes.

There is still plenty open questions regarding usage that focusing on implementation might be to early.

What do you think? :)

File: actionview/lib/action_view/helpers/form_with_helper.rb (hunk 4 above)
Comment (marekkirejczyk) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: Or one can do that:
```
form_with model: @post do |form|
  # Other fields...
  form.checkbox name: 'terms[conditions]', checked: false
end
```

And one can use fields to add to existing scope as a one liner:
fields(:permission) { |f| f.checkbox :admin }

You convinced me. It was in original idea by DHH thou. Shell I remove `scope` related code completely?

File: actionview/lib/action_view/helpers/form_with_helper.rb (hunk 7 above)
Comment (marekkirejczyk) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: This is nice, but doesn't cover pass `value: nil` to force no `value` attribute, e.g. `f.text_field("title", value: nil)`. 

In similar way it doesn't support forcing size not to be shown by forcing nil, i.e. 
`f.text_field("title", maxlength: 35, size: nil)`

One question is: do we want to support old logic that copy `maxlength` to `size` -> might be a bit misleading.

Check tests: `test_text_field_with_nil_value`, `test_text_field_assuming_size`, `test_text_field_removing_size` for more details.

How about:
```
def field_options(field_type, attribute, value = nil, *args, placeholder: nil, **options)
  options = default_options(attribute, options)

  if options.include?(:value)
  elsif value
    options[:value] = value
  elsif @model && !field_type.match('password')
    options[:value] ||= @model.public_send(attribute)
  end

  options[:placeholder] = placeholder(placeholder, attribute)
  options[:size] = options[:maxlength] unless options.key?(:size)
  options[:type] ||= field_type
  options
end
```

Not sure what to do with empty if statement - seems ugly. Any hints?

Also: max_length -> maxlength

File: actionview/lib/action_view/helpers/form_with_helper.rb (hunk 8 above)
Comment (marekkirejczyk) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: Not sure what you mean. Do you meant this string: `'#{selector}'`? It will evaluate during class loading. The generated method will look like this (after including feeback from the other comment; tested in console):

```
def text_field(attribute, *args, **options)                                              
  tag.input field_options('text', attribute, *args, options) 
end

```

**Comment by marekkirejczyk** [comment]
@kaspth Completed work to implement comments from CR, improve style in general and remove unused code. I also added radio buttons and nested fields. 
I am trying not to grow this PR too big. I think functionality wise it is ~80% through. Test wise ~40%.

What could be postpone to next PR: 
- documentation
- tests for more exotic helpers i.e. color_field, date_field etc
- everything related to select helpers family - there is plenty of work there
- custom builders

What I would finish in this PR are form_for tests. Just to make sure all basic cases are covered.

Looking forward to get feedback.

File: actionview/lib/action_view/helpers/form_with_helper.rb (hunk 12 above)
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: > I think it would be easier and more efficient if you start code review from tests, so that we discuss interface and different variants & cases -> then I implement conclusions -> and the we go back to implementation details do the review and implement changes.

As you've probably noticed I'm more of a nuts and bolts guy, preferring to go from the bottom up and to learn decisions from that way. But you're right, let's go the other way — or we might end up with this taking a lot longer.

Thanks for the reminder ❤️

File: actionview/lib/action_view/helpers/form_with_helper.rb (hunk 7 above)
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: Good points 👍

We can use `fetch` to circumvent the empty if:

```ruby
options.fetch(:value) do
  if value
    options[:value] = value
  elsif @model && !field_type.match?('password') # We want to use our Ruby 2.4 backported match? method
    options[:value] = @model.public_send(attribute)
  end
end
```

Yeah, let's find a way to kill either `maxlength` or `size` (also personally don't like `maxlength` as one word).

File: actionview/lib/action_view/helpers/form_with_helper.rb (hunk 4 above)
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: > form.checkbox

Oh, we have wrappers that will generate the correct `input` from that? It ought to be `check_box` though 😉

> And one can use fields to add to existing scope as a one liner:
fields(:permission) { |f| f.checkbox :admin }

Yes, that too!

I only see `:scope` in the `form_with` calls in @dhh's examples and not in individual `*_field` calls.

File: actionview/lib/action_view/helpers/form_with_helper.rb (hunk 8 above)
Comment (kaspth) [inline actionview/lib/action_view/helpers/form_with_helper.rb]: I thought we could pull the string outside `class_eval` such that we could reference that instance from within the method body and spare a new instance each time the method is called. But this isn't really an allocation hot spot, so it's fine 👍

**Comment by kaspth** [comment]
> everything related to select helpers family - there is plenty of work there still

Oh, yeah, `FormOptionsHelper` is going to be a big one.

Only got to respond to your comments today, I'll check back in on the tests soon 😁

**Comment by rafaelfranca** [comment]
I have a feeling that we are undoing all work I did in #4488.

The point of creating reusable classes in #4488 was to make easier to extend existing inputs and create news inputs in both the normal workflow or using form builders like [Simple Form](https://github.com/plataformatec/simple_form). I had a dream to one day make Rails default form builder extensible enough to things like `f.input :name, type: :my_custom_input` possible, where `MyCustomInput` is a class that knows how to render itself.

I believe it is possible and better to the framework at long term to reuse the classes created in #4488. Even more that the problem we are trying to solve is more a DSL problem than an implementation problem.

Right now we have two different DSLs to build forms depending if we have an object or not. Those two different DSLs already shares the same implementation. We are creating a new DSL with a new implementation to solve and we will end up with three DSLs and two implementations.

I'd focus on the problem we want to solve, unify the DSLs. Later we can revisit the defaults of current implementation. And if we manage to create a new implementation that satisfies the extensibility requirements of the current, we can go in that direction.

**Comment by marekkirejczyk** [comment]
@rafaelfranca That was my initial thinking, we have this pretty reusable classes family, why not use them? That is how implementation started. Then I encounter challenges:

1) The implementation is not HTML5 compliant
To solve this we would need to:
- break backward compatibility or
- implement class hierarchy to support two modes - that would lead to code distributed among multiple classes and modules complicating implementation 

2) There is plenty of logic that will have no use in new implementation 
i.e. generating tag id, name, retrieving objects by name
And on the other hand some we would like to have freedom to add/remove/change options.

This logics spans across the whole hierarchy starting from processing arguments to actual logic. 
Solving this would require again one of two: break compatibility or supporting two modes of operations. 

Keeping in mind goals: HTML5 compliance and significant simplification of DSL... how would approach it/continue?

**Comment by rafaelfranca** [comment]
> The implementation is not HTML5 compliant

Why they are not HTML5 compliant? AFAIK they are, and if they are not this is something we can fix and we should fix to all implementations. Like I said the goal of the David's proposal is to unify the DSLs. Anything beyond that can be worked but it is future work.

> There is plenty of logic that will have no use in new implementation 
> i.e. generating tag id, name, retrieving objects by name
> And on the other hand some we would like to have freedom to add/remove/change options.

This is the defaults revisiting part of my coment. The tag id will not be needed but name will. The form needs to know how to build automatic names otherwise there is no point of having a DSL if you have to explicitly set everything.

> Keeping in mind goals: HTML5 compliance and significant simplification of DSL... how would approach it/continue?

By parts.

First, lets focus on the DSL. Having a new DSL building the same HTML code than the actual DSLs.

HTML5 compliance is a bug in the previous implementation and in my opinion and needs to be fixed on it too since, like I said, it is expected to be HTML compliant now.

Defaults revisiting is the last step and if we do the DSL part right, it can be just a matter of configuring the form builder object.

**Comment by marekkirejczyk** [comment]
@rafaelfranca 
Looking at the code level, starting with simple text_field:
```
form_for @post do 
  f.text_field :title
end`
```
the text_field line will translate to `text_field(@post, :title)` and finally to: `Tags::TextField.new(@post, :title, self, {}).render`
 and will produce following html:
`<input type="text" name="post_title" id="post_title" />`

There are two things we would like to change here:

- remove '/' at the end of tag (HTML5 compliance)
- get rid of id

So that we have:
`<input type="text" name="post_title">`

Now how do we pass this information from new `FormWithBuilder` to `Tags::TextField.new` constructor without braking compatibility? 
My first idea was: `Tags::TextField.new(object_name, :title, self, {html5: true, id: false}).render`

But that would generate quite a lot of if-this-then-that code, i.e. in Tags::TextField:
```
if options[:html5]
  tag("input", options)
else
  tag.input(options)
end
```

This same for almost every single Tag and for every single feature. I.e.:
- What if we would like to change options of checkbox (i.e. `check_box :admin, on: 'yes', off: 'no'` instead of `check_box :admin, checked_value: 'yes', unchecked_value: 'no'`), radio button or select? Do we put if in every single class for every single option? Do we build two different implementations in two different classes (`CheckBox` and `NewCheckBox`)? Or build translation logic into new `FromWithBuilder`?
- What if we want to use named parameters in Tags::Base initializers? Can we do that or we have to stick to the current implemenation?

I happy to follow any path you show :) Just let me know how we should approach this design trade-offs.

**Comment by rafaelfranca** [comment]
> There are two things we would like to change here:
> 
> remove '/' at the end of tag (HTML5 compliance)
> get rid of id

Those are exactly what I called defaults revisiting, they don't need to happen now.

The first one is unnecessary, having `/` in the end of the tag is perfectly valid HTML5. See https://www.w3.org/TR/html5/syntax.html#void-elements.

The second one is valid and we don't need to duplicate all the code just to make this happen. Again, lets solve the DSL problem, later we revisit the defaults.

> What if we would like to change options of checkbox (i.e. check_box :admin, on: 'yes', off: 'no' instead of check_box :admin, checked_value: 'yes', unchecked_value: 'no'), radio button or select? Do we put if in every single class for every single option? Do we build two different implementations in two different classes (CheckBox and NewCheckBox)? Or build translation logic into new FromWithBuilder?

If we want this in our new DSL it is easy to archive without changing anything in the Tags implementation.

```
def check_box(name, on: 'yes', off: 'no')
  Tags::Checkbox.new(name, checked_valid: on, unchecked_valid: off)
end
```

> What if we want to use named parameters in Tags::Base initializers? Can we do that or we have to stick to the current implemenation?

This is also not required. Tags is an internal class we don't need to change the interface of the internal class to please our DSLs needs. We can build whatever DSL we want and still use the current interface of the Tags classes.

**Comment by marekkirejczyk** [comment]
Thanks for quick response.
Just to make sure - do I understand correctly? - what you're proposing is:

- keep `/>` at the end of void tags i.e. `<input type="text" />`
- when it comes to change in default Tags::Base classes behavior, we will do minimum of what is necessary, and implement those behavior with additional params and conditions, i.e.
`Tags::TextField.new(object_name, :title, self, {show_id: false}).render`
```
if options[:show_id]
  tag("input", options)
else
  tag.input(options)
end

```
- do the parameters logic translation inside new builder i.e.

```
def check_box(name, on: 'yes', off: 'no')
  Tags::Checkbox.new(name, checked_valid: on, unchecked_valid: off)
end
```
- skip significant changes in Tags::Base hierarchy interface/api to keep backward compatibility?

I understand that those point are important defaults to revisit. Please correct if I misinterpret something. 

I fine to go either way. I guess it is more decision to be made by you, @dhh and @kaspth. I'll be happy to implement whatever makes Rails developers happy :) Just let me know if we're stable on requirements :)

**Comment by kaspth** [comment]
> when it comes to change in default Tags::Base classes behavior, we will do minimum of what is necessary

That decision comes later 😁

> do the parameters logic translation inside new builder i.e

We might not need new builders right now. That was @rafaelfranca's point. This PR should just implement a new DSL using the existing Tags classes.

> skip significant changes in Tags::Base hierarchy interface/api to keep backward compatibility?

We'll see about that when we revisit defaults in a later PR.

Just to reiterate. This PR: add a new DSL that unifies `form_for`/`form_tag` into `form_with`. A later PR: fine tune the HTML that is output and revisit the output defaults.

Let me know if you have any questions 😄

**Comment by kaspth** [comment]
@marekkirejczyk hey 👋, I opened a new PR to get this going again: https://github.com/rails/rails/pull/26976 — credited you as well 😉
